docs/application_architecture.md

```diff
@@ -0,0 +1,104 @@
+## OmiseGO Application Architecture
+
+### Watcher Service
+
+#### Purpose
+
+The watcher first and foremost plays a critical security role in the system.
+
+#### Design principle
+
+- Only include functionality that is critical to the operation of the Plasma security model
+- Strict focus on security role reduces complexity and attack surface area
+- Limited feature helps scalability
+  - The more the watcher does, the slower it can verify
+- 3 primary security functions:
+  - Tracking of the root chain submissions, pulling block contents (from somewhere) and validating, in order to ensure safety of funds passively in possession on the child chain. Watcher notifies in case of the funds are jeopardized.
+  - Proxy API to the child chain API (whatever it may be - PoA server or a P2P PoS network) and the root chain, that ensures that these two are never talked to if the chain is invalid or in unknown state. Only proxy calls that require the chain is operational.
+  - Storage of data critical to access of the funds - UTXO positions, `txbytes` or any other kinds of proofs
+
+#### Requirements
+
+- Events
+  - Should exit due to fault
+    - Double spend
+    - Invalid exit started
+    - Transaction spent
+    - Out of nowhere transaction created
+    - Block withholding?
+    - ...
```

Comment:
@pnowosie – What does the watcher currently detect and what is planned so far?

Reply:
Currently Watcher publishes events in 2 topics
**`transfer:{address}`**
 * `address_received`
 * `address_spent`

**`byzantine`**
 * `block_withholding`
 * `invalid_block`
 * `invalid_exit`
 * `unchallenged_exit`

More details about events and planned ones (WIP) - I've just pushed [branch with `api specs`](https://github.com/omisego/elixir-omg/tree/feature/api-specs-proposal) which @kevsul is taking over